src/ffi/array.rs

```diff
@@ -241,12 +235,21 @@ unsafe fn create_buffer<T: NativeType>(
         return Ok(Buffer::new());
     }
 
-    let ptr = get_buffer_ptr(array, data_type, index)?;
-
     let offset = buffer_offset(array, data_type, index);
-    let bytes = Bytes::from_foreign(ptr, len, BytesAllocator::InternalArrowArray(owner));
+    let ptr: *mut T = get_buffer_ptr(array, data_type, index)?;
 
-    Ok(Buffer::from_bytes(bytes).sliced(offset, len - offset))
+    // We have to check alignment.
+    // This is the zero-copy path.
+    if ptr.align_offset(std::mem::align_of::<T>()) == 0 {
+        let bytes = Bytes::from_foreign(ptr, len, BytesAllocator::InternalArrowArray(owner));
+        Ok(Buffer::from_bytes(bytes).sliced(offset, len - offset))
+    }
+    // This is the path where alignment isn't correct.
+    // We copy the data to a new vec
+    else {
```

Comment:
We copy to a new vec. After this `owner` will go out of scope, calling release on the C FFI, dropping backed memory (if needed).